rate limiting test cases (#2)
psturc/integreatly-operator: test-cases/tests/alerts/c18-validate-api-usage-alerts.md

```
oc get cm rate-limit-alerts -n redhat-managed-api-operator
```

2. Verify that level1, level2 and level3 rate-limiting alerts are present

```
oc get prometheusrules -n redhat-managed-api-marin3r
```

3. Modify the `sku-limits-managed-api-service` configmap to set the rate limit to 20 requests per minute.
   Run

```
oc edit configmap sku-limits-managed-api-service -n redhat-rhmi-operator
```

And insert the following data:

```
data:
  rate_limit: |-
    {
      "RHOAM SERVICE SKU": {
        "unit": "minute",
        "requests_per_unit": 20
      }
    }
```

4. Modify the `rate-limit-alerts` to allow alerts to fire in a reasonable time from the testing perspective:

```
oc edit configmap rate-limit-alerts -n redhat-rhmi-operator
```

and insert the following data:

```
{
      "api-usage-alert-level1": {
        "ruleName": "Level1ThreeScaleApiUsageThresholdExceeded",
        "level": "warning",
        "minRate": "80%",
        "maxRate": "90%",
        "period": "1m"
      },
      "api-usage-alert-level2": {
        "ruleName": "Level2ThreeScaleApiUsageThresholdExceeded",
        "level": "warning",
        "minRate": "90%",
        "maxRate": "95%",

>         "period": "2h"

Reviewer: ```suggestion
        "period": "1m"
```

Reviewer: I'd set the period for the other levels to 1m beforehand as they're going to be tested as well

      },
      "api-usage-alert-level3": {
        "ruleName": "Level3ThreeScaleApiUsageThresholdExceeded",
        "level": "warning",
        "minRate": "95%",

>         "period": "30m"

Reviewer: ```suggestion
        "period": "1m"
```

      }
}
```

> 4. Go to `Networking` > `Routes` under `redhat-rhmi-3scale` namespace

Reviewer 2: @sergioifg94 would you mind reviewing the steps outlined by @psturc and updating where appropriate please?

5. Click on the `zync` route that starts with `https://3scale-admin...`
6. Go to `Secrets` > `system-seed` under 3Scale namespace and copy the admin password
7. Go back to 3Scale login page and login
8. Click on `Ok, how does 3scale work?` and follow the 3Scale wizard to create an API
9. Once on your API overview page, click on `Integration` on the left, then on `Configuration`
10. Copy the `example curl for testing` for `Staging-APIcast` and paste into a terminal window
11. Run the following script to verify that rate limit (20 requests/minute) works correctly:

```

> for i in {1..21}; do <replace-with-example-curl-for-testing>; done

Reviewer: ```suggestion
# Select which number of requests to perform based on the level to trigger
# To trigger level 1 alert
num_req=17
# To trigger level 2 alert
num_req=18
# To trigger level 3 alert
num_req=21
for (( c = 1; c<=$num_req; c++ )); do <replace-with-example-curl-for-testing>; done
```

We need a specific number of requests to trigger level 1 and 2 alerts, as they must be within a range. If we go over the range, only the level 3 alert will fire

```

> > Only the last request should fail (TODO: any specific status code?)

Reviewer: ```suggestion
> Only the last request for level 3 should fail with a `428 Too Many Requests` status code
```

> 12. TODO: should we update the configmap now to increase the rate limit for test the rate-limit alerts?

Reviewer: I think that we can test the 3 levels without modifying the rate limit configuration. I'm still trying to find the sweet spot to replicate the alerts as consistently as possible, so maybe 20 reqs/minute is not the best, but it's possible to replicate all of the levels with that value

Reviewer 2: We could always look to set the dev limits to 100 requests per minute which would make our calculations much easier

13. Run the following command and log in to Prometheus service (as a kubeadmin)

```
open "https://$(oc get routes prometheus-route -n redhat-rhmi-middleware-monitoring-operator -o jsonpath='{.spec.host}')"
```

14. Go to Alerts -> and search for `Level1ThreeScaleApiUsageThresholdExceeded` alert
15. Click on the alert name -> click on the expression (link) -> Graph
16. Run the `curl` command against the APIcast endpoint so you reach between 80%-90% rate limit (i.e. amount of curl requests = rate-limit-value \* 0.85)
17. Verify that the alert graph is showing some data

> 18. Repeat the same process for `Level2ThreeScaleApiUsageThresholdExceeded` and `Level3ThreeScaleApiUsageThresholdExceeded` alerts based on their min and max rates and verify that the alerts are firing as expected

Reviewer: ```suggestion
18. Repeat the same process for `Level2ThreeScaleApiUsageThresholdExceeded` and `Level3ThreeScaleApiUsageThresholdExceeded` alerts based on their min and max rates and verify that the alerts are firing as expected. Make sure to wait at least a minute before running each set of `curl` commands
```

If we run the set of requests too soon, the new requests will be accumulated to the previous one, reaching the rate limit earlier than expected

---
products:
  - name: rhoam
    environments:
      - osd-fresh-install
estimate: 1h
tags:
  - per-release
---

# C18 - Validate API usage alerts

## Prerequisites

1. Login via `oc` as a user with **cluster-admin** role (kubeadmin):

```
oc login --token=<TOKEN> --server=https://api.<CLUSTER_NAME>.s1.devshift.org:6443
```

2. RHOAM is installed

## Steps

1. Validate that rate-limiting alerts ConfigMap has been created

```
oc get cm rate-limit-alerts -n redhat-managed-api-operator
```

2. Verify that level1, level2 and level3 rate-limiting alerts are present

```
oc get prometheusrules -n redhat-managed-api-marin3r
```

> 3. (TODO: modify the alerts configmap - configmap name?)

Reviewer 2: Assuming that this test case is performed on a cluster where RHOAM has not been installed via the add-on, we should be able to modify the `sku-limits-managed-api-service` configmap in the RHOAM operator namespace to something like:
```
data:
  rate_limit: |-
    {
      "RHOAM SERVICE SKU": {
        "unit": "minute",
        "requests_per_unit": 20
      }
    }
```
Once the rate limit configs are updated, we should then create a new 3Scale application by logging into 3scale-admin route (login creds can be retrieved from the system-seed secret) and follow the getting started wizard. This provides us with an endpoint that will be rate limited.

Useful command for running X number of requests per second:
```
for i in {1..X}; do curl -i https://api-3scale-apicast-staging.apps.<cluster-id>.devshift.org//?user_key=XXXX; done
```
@sergioifg94 might should have detailed steps on how to trigger each alert

Reviewer 2: It seems that the steps for creating a new 3Scale application are listed in `h22-validate-that-rate-limit-service-is-working-as-expected.md` so we should be able to copy/paste them

Reviewer: For now I've only managed to trigger the level 3 one, but I'm gonna try to play with the different configurations to get the other ones

Author: Thanks for the additional info guys.

Regarding checking the alerts - maybe this script that resiliency team is using for checking the alerts during performance testing could help us to easily detect the alerts?
https://github.com/integr8ly/integreatly-operator/blob/master/scripts/alerts-during-perf-testing.sh
